pirouette-plutusir/src/Language/Pirouette/PlutusIR/SMT.hs

```diff
@@ -0,0 +1,297 @@
+              [ Branch {additionalInfo = And [Native rel], newTerm = reifyBoolean True},
+                Branch {additionalInfo = And [Native $ PureSMT.not rel], newTerm = reifyBoolean False}
+              ]
+  branchesBuiltinTerm op _ [TermArg (K (PIRConstByteString x)), TermArg (K (PIRConstByteString y))]
+    | op `elem` plutusIRBasicOps || op `elem` plutusIRBasicRels =
+      (\r -> pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]) $
+        case op of
+          AppendByteString -> K $ PIRConstByteString (x <> y)
+          EqualsByteString -> reifyBoolean $ x == y
+          LessThanByteString -> reifyBoolean $ x < y
+          LessThanEqualsByteString -> reifyBoolean $ x <= y
+          _ -> error "ill-typed application"
+  branchesBuiltinTerm LengthOfByteString _ [TermArg (K (PIRConstByteString b))] =
+    let r = K $ PIRConstInteger (fromIntegral $ BS.length b)
+     in pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]
+  branchesBuiltinTerm
+    ConsByteString
+    _
+    [TermArg (K (PIRConstInteger i)), TermArg (K (PIRConstByteString b))] =
+      let r = K $ PIRConstByteString (BS.cons (fromInteger i) b)
+       in pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]
+  branchesBuiltinTerm
+    IndexByteString
+    _
+    [TermArg (K (PIRConstByteString b)), TermArg (K (PIRConstInteger i))] =
+      let r =
+            if i < 0
+              then errorTerm
+              else K $ PIRConstInteger $ fromIntegral (BS.index b (fromInteger i))
+       in pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]
+  branchesBuiltinTerm
+    IndexByteString
+    _
+    [TermArg (K (PIRConstInteger start)), TermArg (K (PIRConstInteger n)), TermArg (K (PIRConstByteString xs))] =
+      let r = K $ PIRConstByteString $ BS.take (fromInteger n) (BS.drop (fromInteger start) xs)
+       in pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]
+  branchesBuiltinTerm op _ [TermArg (K (PIRConstString x)), TermArg (K (PIRConstString y))]
+    | op `elem` plutusIRBasicOps || op `elem` plutusIRBasicRels =
+      (\r -> pure $ Just [Branch {additionalInfo = mempty, newTerm = r}]) $
+        case op of
+          AppendString -> K $ PIRConstString $ x <> y
+          EqualsString -> reifyBoolean $ x == y
+          _ -> error "ill-typed application"
+  -- if-then-else goes to the helpers
+  -- branchesBuiltinTerm IfThenElse _ (TyArg _ : TermArg c : TermArg t : TermArg e : excess) =
+  --   let isEq EqualsInteger = True
+  --       isEq EqualsString = True
+  --       isEq EqualsByteString = True
+  --       isEq _ = False
+  --       isTrue (K (PIRConstBool True)) = True
+  --       isTrue _ = False
+  --       isFalse (K (PIRConstBool False)) = True
+  --       isFalse _ = False
+  --    in ifThenElseBranching
+  --         isTrue
+  --         (K (PIRConstBool True))
+  --         isFalse
+  --         (K (PIRConstBool False))
+  --         isEq
+  --         c
+  --         t
+  --         e
+  --         excess
```

Comment:
Will we need this code in the future?

Reply:
that's a good question! I don't know tbh